templates/template.go

```diff
@@ -0,0 +1,299 @@
+	copy(t.Params, base)
+	for _, p := range currentParams {
+		if i, item := t.paramWithName(p.Name); item != nil {
+			// we only allow overwriting a few fields
+			if p.Default != "" {
+				t.Params[i].Default = p.Default
+			}
+			if p.Example != "" {
+				t.Params[i].Example = p.Example
+			}
+		} else {
+			t.Params = append(t.Params, p)
+		}
+	}
+}
+
+// Defaults returns a map of default values for the template
+func (t *Template) Defaults(docsOrTests bool) map[string]interface{} {
+	values := make(map[string]interface{})
+	for _, p := range t.Params {
+		if p.Test != "" {
+			values[p.Name] = p.Test
+		} else if p.Example != "" && docsOrTests {
+			values[p.Name] = p.Example
+		} else {
+			values[p.Name] = p.Default // may be empty
+		}
+	}
+
+	return values
+}
+
+// return the param with the given name
+func (t *Template) paramWithName(name string) (int, *Param) {
+	for i, p := range t.Params {
+		if p.Name == name {
+			return i, &p
+		}
+	}
+	return 0, nil
+}
+
+// Usages returns the list of supported usages
+func (t *Template) Usages() []string {
+	if _, p := t.paramWithName(ParamUsage); p != nil {
+		return p.Choice
+	}
+
+	return nil
+}
+
+func (t *Template) ModbusChoices() []string {
+	if _, p := t.paramWithName(ParamModbus); p != nil {
+		return p.Choice
+	}
+
+	return nil
+}
+
+//go:embed proxy.tpl
+var proxyTmpl string
+
+// RenderProxy renders the proxy template for inclusion in documentation
+func (t *Template) RenderProxy() ([]byte, error) {
+	return t.RenderProxyWithValues(nil, false)
+}
+
+func (t *Template) RenderProxyWithValues(values map[string]interface{}, includeDescription bool) ([]byte, error) {
+	tmpl, err := template.New("yaml").Funcs(template.FuncMap(sprig.FuncMap())).Parse(proxyTmpl)
+	if err != nil {
+		panic(err)
+	}
+
+	for index, p := range t.Params {
+		for k, v := range values {
+			if p.Name == k {
+				t.Params[index].Value = v.(string)
+			}
+		}
+	}
+
+	// remove params with no values, no defaults and no example
+	var newParams []Param
+	for _, param := range t.Params {
+		if param.Value == "" && param.Default == "" && param.Example == "" && !param.Required {
+			continue
+		}
+		newParams = append(newParams, param)
+	}
+
+	for index, p := range newParams {
+		newParams[index].Value = yamlQuote(p.Value)
+	}
+
+	t.Params = newParams
+
+	out := new(bytes.Buffer)
+	data := map[string]interface{}{
+		"Template": t.Template,
+		"Params":   t.Params,
+	}
+	if includeDescription {
+		data["Description"] = t.Description
+	}
+	err = tmpl.Execute(out, data)
+
+	return bytes.TrimSpace(out.Bytes()), err
+}
+
+// RenderResult renders the result template to instantiate the proxy
+func (t *Template) RenderResult(docs bool, other map[string]interface{}) ([]byte, map[string]interface{}, error) {
+	values := t.Defaults(docs)
+	if err := util.DecodeOther(other, &values); err != nil {
+		return nil, values, err
+	}
+
+	values = t.ModbusValues(values)
+
+	for item, p := range values {
+		values[item] = yamlQuote(p.(string))
```

Comment:
Hier knallts wenn in einer Template-Config im yaml eine Zahl steht.

Beispiel:
```
vehicles:
- type: template
  template: tesla
  title: foobar
  capacity: 82     <<<
```

oder

```
meters:
- type: template
  template: vzlogger_http
  usage: grid
  host: 192.168.31.53
  port: 9000             <<<<
```

Reply:
Danke, gefixt.

Reply:
Klappt. Danke!